3 단계 - 자동차 경주

hyyg123 opened the pull request:
2단계 문자열 계산기, 피드백에 따라 리팩토링 진행하였습니다.
- 수식, 연산자, 피연산자 클래스 추가
- 문자열 분리 정책, 연산자/피연산자 추출 정책 추가
- 패키지 수정

3단계 자동차 경주 구현
- 2단계보다 도메인 분리, 계증 분리에 좀 더 신경썼습니다.
- IntRange 를 처음으로 사용해봤습니다.

훌륭한 리뷰 항상 감사합니다 :)

Commit:
refactor : Expression, Operand, Extract

리팩토링 임시

hyyg123 committed:
refactor : 역할 및 책임 분리

- 수식, 연산자, 피연산자 객체 추가
- 문자열 분리 정책, 연산자/피연산자 추출 정책 분리
- 패키지 구분

hyyg123 committed:
docs : 3단계 자동차 경주 요구사항 작성

- README, 요구사항 추가 작성

hyyg123 committed:
feat : 3단계 자동차 랜덤 경주 구현

- Domain Layer, Service Layer, Presentation Layer 구분하여 설계

hyyg123 committed:
refactor : 2단계 문자열 계산기 리팩토링

- import wildcard 제거

hyyg123 committed:
refactor : 3단계 자동차 경주 리팩토링

- ErrorType 추가
- for loop 대신 sequence 이용
- 값 검증 로직 추가
- 값 검증 실패 케이스 테스트 추가

hyyg123 committed:
refactor : Random.nextInt 수정

- from 값 제거

hyyg123 committed:
refactor : ExpressionExtractor extractPolicy 인터페이스화

- 인터페이스 타입으로 받도록 수정

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCars.kt:
```diff
@@ -0,0 +1,29 @@
+
+import racingcar.contants.ErrorType
+
+class RacingCars(
```
일급 컬렉션 👍

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCars.kt:
```diff
@@ -0,0 +1,29 @@
+            ErrorType.INVALID_NUMBER.message
+        }
+
+        racingCars = (1..racingCarNum)
+            .asSequence()
+            .map { RacingCar(movePolicy) }
+            .toList()
```
```suggestion
        racingCars = List(racingCarNum) {
            RacingCar(movePolicy)
        }
```

hyyg123 replied:
오!! 이런 게 있었군요! 😀

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -0,0 +1,13 @@
+package racingcar.domain
+
+class RacingCar(
```
해당 클래스는 data class를 활용해봐도 좋았을 것 같네요

hyyg123 replied:
RacingCar 클래스를 데이터 클래스로 활용했을 때 장점이 있을까요?

현재 RacingCar 클래스가 가지고 있는 멤버 필드는 movedDistance, movePolicy 인데,
해당 값이 동일하다고 해서 동일한 객체로 판단하는 게 맞는지 의문이어서요!
(toString(), hashCode(), equals(), copy() 메소드들이 꼭 필요할지 ?)

lhs0610 replied:
음 그렇네요 ㅎㅎ equals override를 고려했던건데 굳이 필요가 없겠네요. 해당 피드백은 패스해도 될 것 같습니다

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCar.kt:
```diff
@@ -0,0 +1,13 @@
+class RacingCar(
+    private val movePolicy: MovePolicy
+) {
+    var movedDistance: Int = 0
```
여기서 좀 더 고민해보면 일급 객체로 포장할수도 있을 것 같아요 👍

lhs0610 commented on src/main/kotlin/racingcar/domain/RandomMovePolicy.kt:
```diff
@@ -0,0 +1,20 @@
+
+import kotlin.random.Random
+
+class RandomMovePolicy : MovePolicy {
```
해당 전략은 별도의 상태를 가지지 않는 무상태(stateless) 객체로 보이는데, class가 아니라 object로 싱글턴으로 만들어보면 어떨까요?

lhs0610 commented on src/main/kotlin/racingcar/service/Race.kt:
```diff
@@ -0,0 +1,33 @@
+import racingcar.model.RaceResultDto
+import racingcar.model.StepResultDto
+
+class Race(
+    racingCarNum: Int,
+    private val stepNum: Int
+) {
```
서비스 계층인데 상태를 가지도록 설계가 되어 있네요 🤔
서비스 계층은 무상태로 설계해서 재사용성읖 놀힐 수 있을 것 같은데 어떻게 생각하시나요?

hyyg123 replied:
그렇네요! start 함수에 stepNum 파라미터를 받아서 처리해보겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/service/Race.kt:
```diff
@@ -0,0 +1,33 @@
+            ErrorType.INVALID_NUMBER.message
+        }
+
+        racingCars = RacingCars(racingCarNum = racingCarNum, movePolicy = RandomMovePolicy())
```
객체가 특정 전략 구현체에 의존하고있는데DIP원칙을 위반하는 것 같은데 어떻게 생각하시나요?

hyyg123 replied:
그렇네요! 외부에서 movePolicy 주입하도록 수정하겠습니다.

lhs0610 commented on src/main/kotlin/racingcar/ui/ResultView.kt:
```diff
@@ -0,0 +1,14 @@
+    println("실행 결과")
+    raceResult.stepResults.forEach { stepResult ->
+        stepResult.movedDistanceList.forEach { movedDistance ->
+            (1 .. movedDistance).forEach { _ -> print("-") }
```
- 같은 매직리터럴 역시 상수로 분리해주면 어떨까요

lhs0610 commented on src/main/kotlin/racingcar/ui/ResultView.kt:
```diff
@@ -0,0 +1,14 @@
+
+fun printRaceResult(raceResult: RaceResultDto) {
+    println("실행 결과")
+    raceResult.stepResults.forEach { stepResult ->
```
현재 for문이 2중 첩되어 indent가 3을 넘어가는데 어떻게 줄여볼 수 없을까요? 
또한 각각의 스키드마크에서 매번 출력을 호출하면 성능적으로도 부하가 걸릴 수 있을 것 같아요. 
StringBuilder를 사용해볼 수 없을까요? 코틀린에서는 buildString같은 DSL을 제공하고 있습니다.

lhs0610 commented on src/main/kotlin/racingcar/ui/ResultView.kt:
```diff
@@ -0,0 +1,14 @@
+    println("실행 결과")
+    raceResult.stepResults.forEach { stepResult ->
+        stepResult.movedDistanceList.forEach { movedDistance ->
+            (1 .. movedDistance).forEach { _ -> print("-") }
```
String의 repeat 함수를 사용해보면 어떨까요? 좀 더 간결하게 작성할 수 있을 것 같아요 ㅎㅎ

hyyg123 replied:
오 이런 편의 기능이..!

lhs0610 commented on src/test/kotlin/racingcar/domain/RacingCarsTest.kt:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    Given("자동차 대수 0을 제공한다.") {
+        val racingCarNum = 0
+        When("RacingCars를 만들 때") {
+            Then("예외가 발생한다.") {
+                val exception = shouldThrow<IllegalArgumentException> { RacingCars(racingCarNum) { 1 } }
+                exception.message should startWith(ErrorType.INVALID_NUMBER.message)
+            }
+        }
+
+    }
+
+    Given("자동차 대수 음수를 제공한다.") {
+        val racingCarNum = -1
+        When("RacingCar를 만들 때") {
+            Then("예외가 발생한다.") {
+                val exception = shouldThrow<IllegalArgumentException> { RacingCars(racingCarNum) { 1 } }
+                exception.message should startWith(ErrorType.INVALID_NUMBER.message)
+            }
+        }
+
+    }
```
둘 다 경계값으로 실패하는 테스트인데, 하나로 묶어볼 수도 있을 것 같네요 
```kotlin
Given("유효하지 않은 자동차 대수 크기를 제공한다") {
    listOf(-1, 0).forEach {
        When("RacingCars를 만들 때") { racingCarNum->
            Then("예외가 발생한다.") {
                val exception = shouldThrow<IllegalArgumentException> { RacingCars(racingCarNum) { 1 } }
                exception.message should startWith(ErrorType.INVALID_NUMBER.message)
            }
        }
    }
}

```

lhs0610 reviewed (changes requested):
용구님, 3단계 미션 잘 구현해주셨습니다 👍 
몇 가지 피드백 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요. 

추가적으로 현재 klint 체크가 실패하고있는데 이 부분도 같이 확인해주시면 좋을 것 같습니다.
![image](https://github.com/lhs0610/kotlin-racingcar/assets/138735487/483406f4-b6ee-4781-b3e7-877f920d6596)

Commit:
refactor : 피드백 반영

- Race 서비스에 RacingCars 외부 주입 (with MovePolicy)
- Race 서비스에 stepNum 멤버 변수 제거 -> 메소드 파라미터로 변경
- RandomMovePolicy 클래스, stateless 상태로 싱글톤으로 변경(object)

Commit:
test : 유효하지 않은 자동채 대수 테스트 수정

- 두 테스트 모두 경계값 테스트로 하나의 테스트로 함침

Commit:
refactor : 실행 결과 로직 리팩토링

- 3중 for문 제거 -> joinToString, String.repeat 이용하여 리팩토링

Commit:
refactor : 레이싱 카 이동거리 일급 객체로 리팩토링

- RacingCarDistance 일급 객체 추가

Commit:
refactor : ktlint formatting

- klintFormat 실행

hyyg123 commented:
피드백 반영하여 수정하였습니다.

수정 사항은 아래와 같습니다.
- Race > RacingCars 외부 주입 (with MovePolicy)
- Race > stepNum 멤버 변수 제거 -> 메소드 파라미터로 변경
- RandomMovePolicy > stateless 상태로 싱글톤으로 변경(object)
- RacingCarTest > 경계값 테스트들은 하나의 테스트로 합침
- ResultView > 3중 for loop 제거 -> joinToString, String.repeat 이용하여 리팩토링
- RacingCarDistance 일급 객체 추가
- ktlint formatting

꼼꼼한 리뷰 항상 감사드려요 ~ !
@lhs0610

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCarDistance.kt:
```diff
@@ -0,0 +1,20 @@
+
+import racingcar.contants.ErrorType
+
+class RacingCarDistance(
+    private var distanceValue: Int = 0
+) {
```
이렇게 값이 하나뿐인 값 객체는 다음과 같이 값 객체로 최적화가 가능합니다. 
```suggestion
@JvmInline
value class RacingCarDistance(
    private var distanceValue: Int = 0
) {
```
이렇게 표현하면, 컴파일시점에 해당 객체를 일반 필드처럼 인라이닝해서 객체 참조 오버헤드를 줄일 수 있습니다

Commit:
refactor : IntRange -> List

- IntRange -> List

lhs0610 commented on src/main/kotlin/racingcar/domain/RacingCarDistance.kt:
```diff
@@ -0,0 +1,20 @@
+        }
+    }
+
+    fun increase(moveDistance: Int = 0) {
+        distanceValue += moveDistance
+    }
```
사실 이렇게 구현을 할 경우 내부 필드 변경의 위험이 있어 값 객체는 동시성을 고려하는 Atomic 객체가 아니라면 불변으로 만들어서 관리해주는걸 좀 더 권장하는편이에요 ㅎㅎ 반면 동시성을 고려한다면 내부에 동기화 블럭등을 활용해줄수도 있겠죠!

lhs0610 commented on src/main/kotlin/racingcar/service/Race.kt:
```diff
@@ -2,32 +2,22 @@ package racingcar.service
-    racingCarNum: Int,
-    private val stepNum: Int
-) {
     private val racingCars: RacingCars
```
이 내부 프로퍼티도 제거하는게 맞지 않을까요 🤔

lhs0610 reviewed (approved):
용구님 3단계 미션 피드백 반영까지 빠르게 잘 해주셨어요. 
몇 가지 피드백 남기긴했는데 4단계 진행하시면서 같이 고민해보면 좋을 것 같아요. 4단계는 여기서 merge하도록 하겠습니다!
